File: deck/typescript/code/string/sound.ts
// Talk to sounds.

import { combine } from './combine'
import { codeOf } from '../space/codec'
import { R, modifierAttaches, pickModifier } from './runtime'
import { NO_CODE } from './type'

/**
 * `machine()` reports the tone type at full detail, which is what the
 * flat code always meant.
 */
const TONE = 'tone' as const
const MESH = 'mesh' as const
import type { Modifier, Phone, Sound, SymbolEntry } from './type'

export function makeSound(
  base: Phone,
  mods: Modifier[],
  pre: Modifier[] = [],
): Sound {
  const talk = combine(base.talk, mods, pre)
  const ordered = [...mods].sort((a, b) => a.order - b.order)
  const leading = [...pre].sort((a, b) => b.order - a.order)

  const prefix = ordered.filter(m => m.prefix)
  const suffix = ordered.filter(m => !m.prefix)

  const ipa =
    leading.map(m => m.ipa).join('') +
    prefix.map(m => m.ipa).join('') +
    base.ipa +
    suffix.map(m => m.ipa).join('')

  const simple =
    leading.map(m => m.simple).join('') +
    prefix.map(m => m.simple).join('') +
    base.simple +
    suffix.map(m => m.simple).join('')

  return {
    talk,
    ipa,
    simple,
    machine: codeOf({
      sound: { base, modifiers: ordered },
      type: TONE,
      system: MESH,
    }),
    kind: base.form,
    base,
    modifiers: ordered,
    pre: leading,
    raw: false,
  }
}

function rawSound(entry: SymbolEntry): Sound {
  return {
    talk: entry.talk,
    ipa: entry.ipa,
    simple: entry.simple,
    // A passthrough symbol is not a sound, so it has no code.
    machine: NO_CODE,
    kind: 'symbol',
    modifiers: [],
    pre: [],
    raw: true,
  }
}

// Split a talk string into sounds. A single starter lookup gives the base
// (or a symbol); a base then swallows the modifiers that follow it, and
// the sound is re-emitted in canonical order.
/**
 * The modifiers inside one pair of brackets, or null when the contents are
 * not a clean run of them.
 *
 * ORDER IS FREE ON THE WAY IN. `combine` writes the canonical order, so a
 * stored sound has one spelling, but a person typing `k<hw>` and a person
 * typing `k<wh>` mean the same thing and both should parse. This is the same
 * contract the IPA side already has, where `n̪̥` and `n̥̪` both read and one
 * comes back.
 *
 * NULL RATHER THAN A PARTIAL READ. Anything unrecognised inside the brackets
 * means the whole run is not talk, and the caller carries the characters
 * through untouched instead of keeping the half it understood.
 */

/**
 * Read a bracketed modifier run.
 *
 * `base` is the phone the run FOLLOWS, and is given only in that position.
 * A run written after a base is claiming those marks belong to it, and a
 * mark whose `attaches` rule rules that out means the claim is wrong: the
 * run is really the leading run of whatever comes next. Returning null says
 * so, and the caller falls through to the leading reading.
 *
 * `aʰk` and `mʰk` both spell as `X<h>k`, so the spelling alone cannot say
 * which sound the mark belongs to. Aspiration attaches to a plosive and not
 * to a vowel or a nasal, which is what decides it. In the LEADING position
 * there is nothing to decide, so no base is passed and the rule is not
 * applied.
 */

function readRun(
  run: string,
  form: 'consonant' | 'vowel',
  base?: Phone,
  loose = false,
): Modifier[] | null {
  if (!run) {
    return null
  }

  const out: Modifier[] = []
  let at = 0

  while (at < run.length) {
    const options = R.talkModifier.matchAt(run, at)

    if (options === undefined) {
      return null
    }

    // LOOSE IS THE LAST READING BEFORE GIVING UP. A modifier is normally
    // read against the form of its base, but the encoder writes marks that
    // have no reading for that form: `iʲ` comes back as `i<y^>`, and
    // palatalization is spelled for a consonant. Refusing left the bracket
    // to leak out of the tokenizer one raw character at a time, which is
    // strictly worse than reading it as what it plainly says.
    const one = loose
      ? (pickModifier(options, form) ?? options[0])
      : pickModifier(options, form)

    if (one === undefined) {
      return null
    }

    if (base && !modifierAttaches(base, one)) {
      return null
    }

    out.push(one)
    at += R.talkModifier.matchedLength
  }

  return out.length ? out : null
}

/**
 * The span of a bracketed modifier run starting at `at`, or null.
 *
 * A SCAN, NOT A SEARCH. This read `indexOf('>')`, which finds a closing
 * bracket anywhere ahead: a stray `<` with its match far downstream would
 * swallow every sound between them and hand the contents to the modifier
 * reader as one run. Walking forward instead stops at the first character
 * that cannot be inside a run, so an unclosed `<` fails here and is carried
 * through as an ordinary character rather than eating the rest of the word.
 *
 * Runs do not nest. A `<` inside one is malformed, not a sub-run, so it ends
 * the attempt rather than opening a second.
 */

function runSpan(text: string, at: number): { body: string; end: number } | null {
  if (text[at] !== '<') {
    return null
  }

  let index = at + 1

  while (index < text.length) {
    const one = text[index]

    if (one === '>') {
      return index === at + 1
        ? null
        : { body: text.slice(at + 1, index), end: index + 1 }
    }

    if (one === '<') {
      return null
    }

    index += 1
  }

  return null
}

/**
 * A binder: `<B>` ties the two graphemes before it into one segment, `<B3>`
 * the three before it, and so on.
 *
 * WHY IT COUNTS BACKWARD. A tie in IPA sits BETWEEN the letters it joins,
 * so `t͡ʃ` says nothing about how far the binding reaches: a reader infers
 * two from the position. That works on paper and not in a string a machine
 * segments, which is why the old spelling carried the tie as an opaque
 * symbol standing between two sounds and could not say what it bound.
 *
 * Naming the count instead makes a doubly-articulated `k͡p` and a
 * three-part cluster equally sayable, and it puts the binder after its
 * material, where the modifier brackets already are.
 *
 * MODIFIERS COUNT AS PART OF THEIR SOUND. `t<h>x<B>` binds the aspirated
 * `t` and the `x`, two graphemes, because a mark belongs to the letter it
 * sits on rather than standing beside it.
 */

const BINDER = /^B(\d*)$/

/** How many sounds a binder joins, or null when the run is not a binder. */

function binderReach(run: string): number | null {
  const found = BINDER.exec(run)

  if (!found) {
    return null
  }

  const reach = found[1] ? Number(found[1]) : 2

  // Binding one sound to nothing is not a tie, and a count that runs past
  // the start of the word cannot be honoured either. Both are left for the
  // caller to reject rather than silently clamped.
  return reach >= 2 ? reach : null
}

/**
 * Several sounds joined into one, as a tie does.
 *
 * THE TIE BINDS THE BASES AND THE MARKS HOIST OUT. `t<h>x<B>` is an
 * aspirated affricate and comes back `t͡ʃʰ`, not `tʰ͡ʃ`. A tie says two
 * LETTERS are one segment, so anything between them breaks the very thing
 * it is asserting, and the aspiration belongs to the whole affricate
 * rather than to the stop half of it.
 *
 * `tʰ͡ʃ` still PARSES, because a source may well write it and refusing to
 * read a source is worse than reading it and saying so canonically. It
 * simply comes back in the other order.
 *
 * The result is raw rather than a phone: the parts keep their identities
 * and the binding is a claim about them, not a new entry in the catalog.
 */

function bindSounds(parts: Sound[]): Sound {
  const talk = parts.map(one => one.talk).join('')
  const simple = parts.map(one => one.simple).join('')

  // A part with no base is a passthrough, and its own spelling is all there
  // is to bind; otherwise the base ties and the marks follow the group.
  const bases = parts.map(one => one.base?.ipa ?? one.ipa)
  const marks = parts.flatMap(one => one.modifiers.map(mark => mark.ipa))
  const pre = parts.flatMap(one => one.pre.map(mark => mark.ipa))

  const ipa = pre.join('') + bases.join('\u{0361}') + marks.join('')

  return rawSound({
    talk: `${talk}<B${parts.length > 2 ? parts.length : ''}>`,
    ipa,
    simple,
  })
}

export function segment(text: string): Sound[] {
  const sounds: Sound[] = []

  // Modifiers seen before a base, which modify what FOLLOWS: `h~k` is
  // pre-aspirated, `n~d` prenasalized. No affix spelling is also a starter
  // spelling, so a modifier match here is unambiguous; where both match,
  // the longer one wins, as everywhere else in the scan.
  let leading: Modifier[] = []

  let i = 0

  while (i < text.length) {
    const start = R.talkStarter.matchAt(text, i)
    const startLength = start === undefined ? 0 : R.talkStarter.matchedLength

    // A BRACKETED RUN BEFORE A BASE is the pre-modifiers, `<h>k` for
    // pre-aspirated. Read whole, then split into tokens, so the scan below
    // never sees a `<` or a `>`.
    if (startLength === 0) {
      const span = runSpan(text, i)

      // A BINDER TIES WHAT CAME BEFORE IT, so it is read here rather than
      // as a modifier on a following base: `tx<B>` is one affricate, and
      // there is no base after the bracket to carry it.
      if (span) {
        const reach = binderReach(span.body)

        if (reach !== null && sounds.length >= reach) {
          const held = sounds.splice(sounds.length - reach, reach)

          sounds.push(bindSounds(held))
          i = span.end
          continue
        }
      }

      if (span) {
        const next = R.talkStarter.matchAt(text, span.end)

        if (next !== undefined && next.role === 'phone') {
          const held =
            readRun(span.body, next.phone.form) ??
            readRun(span.body, next.phone.form, undefined, true)

          if (held) {
            leading.push(...held)
            i = span.end
            continue
          }
        }
      }
    }

    if (start === undefined) {
      // Unknown character: carry it through so nothing is silently dropped.
      const ch = text[i++]

      sounds.push(rawSound({ talk: ch, ipa: ch, simple: ch }))
      continue
    }

    // `startLength` was captured BEFORE the lookahead below, because
    // `matchAt` stores its result on the trie: probing for a base after a
    // candidate pre-modifier overwrites `matchedLength`, and reading it
    // here would advance by the wrong amount, sometimes by zero.
    i += startLength

    if (start.role === 'phone') {
      const mods: Modifier[] = []

      // A BRACKETED RUN AFTER A BASE carries every modifier on it, in one
      // pair. `k<wh>` is labialized and aspirated. The contents are
      // uniquely decodable, so no separator is needed inside.
      const span = runSpan(text, i)

      if (span) {
        // WHOSE MARK IS IT. A run written after a base usually belongs to
        // it, and the `attaches` rule is what settles the cases where it
        // might not: aspiration goes to a plosive, not to a vowel.
        //
        // But the rule alone is too strong. `ipaToTalk` spells b̥ as
        // `b<v->` and kǃʰ as `k!<h>`, and neither voicelessness nor
        // aspiration attaches to those bases by rule, so refusing outright
        // meant the tokenizer could not read back what its own encoder
        // writes and the brackets spilled out as raw characters.
        //
        // A mark is only handed on when there is somewhere for it to GO:
        // the next sound has to be a phone that accepts the whole run. If
        // nothing ahead will take it, it stays where it was written, which
        // is also the reading that loses nothing.
        const fits = readRun(span.body, start.phone.form, start.phone)
        const held =
          fits ??
          readRun(span.body, start.phone.form) ??
          readRun(span.body, start.phone.form, undefined, true)

        let handOn = false

        if (held && !fits) {
          const next = R.talkStarter.matchAt(text, span.end)

          handOn =
            next !== undefined &&
            next.role === 'phone' &&
            readRun(span.body, next.phone.form, next.phone) !== null
        }

        if (held && !handOn) {
          sounds.push(makeSound(start.phone, held, leading))
          leading = []
          i = span.end
          continue
        }
      }

      sounds.push(makeSound(start.phone, mods, leading))
      leading = []
    } else if (start.role === 'symbol') {
      sounds.push(rawSound(start.symbol))
    }
  }

  // A pre-modifier with nothing after it modifies nothing. Carry the
  // spelling through rather than dropping it, so a caller sees the input
  // was incomplete instead of losing it silently.
  for (const modifier of leading) {
    sounds.push(
      rawSound({
        talk: modifier.talk,
        ipa: modifier.ipa,
        simple: modifier.simple,
      }),
    )
  }

  return sounds
}

File: deck/rust/code/string/sound.rs
//! Talk to sounds.

use crate::space::codec::code_of;
use crate::space::model::{Notation, Tier};
use crate::string::combine::combine;
use crate::string::runtime::{modifier_attaches, pick_modifier, runtime};
use crate::string::types::{NO_CODE, Form, Kind, Modifier, Phone, Sound, SymbolEntry, Unit};

pub fn make_sound(
  base: &'static Phone,
  mods: Vec<&'static Modifier>,
  pre: Vec<&'static Modifier>,
) -> Sound {
  let talk = combine(&base.talk, &mods, &pre);

  let mut ordered = mods;

  ordered.sort_by_key(|modifier| modifier.order);

  let mut leading = pre;

  leading.sort_by_key(|modifier| std::cmp::Reverse(modifier.order));

  let lead = |of: fn(&Modifier) -> &str| -> String {
    leading.iter().map(|modifier| of(modifier)).collect()
  };

  let spell = |pick: fn(&Modifier) -> bool, of: fn(&Modifier) -> &str| -> String {
    ordered
      .iter()
      .filter(|modifier| pick(modifier))
      .map(|modifier| of(modifier))
      .collect()
  };

  let is_prefix: fn(&Modifier) -> bool = |modifier| modifier.prefix;
  let is_suffix: fn(&Modifier) -> bool = |modifier| !modifier.prefix;

  let ipa = format!(
    "{}{}{}{}",
    lead(|m| &m.ipa),
    spell(is_prefix, |m| &m.ipa),
    base.ipa,
    spell(is_suffix, |m| &m.ipa)
  );

  let simple = format!(
    "{}{}{}{}",
    lead(|m| &m.simple),
    spell(is_prefix, |m| &m.simple),
    base.simple,
    spell(is_suffix, |m| &m.simple)
  );

  Sound {
    machine: code_of(base, &ordered, TONE, MESH),
    talk,
    ipa,
    simple,
    kind: base.form.into(),
    base: Some(base),
    modifiers: ordered,
    pre: leading,
    raw: false,
  }
}

fn raw_sound(entry: &SymbolEntry) -> Sound {
  Sound {
    talk: entry.talk.clone(),
    ipa: entry.ipa.clone(),
    simple: entry.simple.clone(),
    // A passthrough symbol is not a sound, so it has no code.
    machine: NO_CODE,
    kind: Kind::Symbol,
    base: None,
    modifiers: Vec::new(),
    pre: Vec::new(),
    raw: true,
  }
}

/// `machine()` reports the tone notation at full detail, which is what the
/// flat code always meant.
const TONE: Notation = Notation::Tone;
const MESH: Tier = Tier::Mesh;

/// The modifiers inside one pair of brackets, or `None` when the contents are
/// not a clean run of them.
///
/// ORDER IS FREE ON THE WAY IN. `combine` writes the canonical order, so a
/// stored sound has one spelling, but a person typing `k<hw>` and a person
/// typing `k<wh>` mean the same thing and both should parse.
///
/// `base` is the phone the run FOLLOWS, and is given only in that position. A
/// run written after a base is claiming those marks belong to it, and a mark
/// whose `attaches` rule rules that out means the claim is wrong: the run is
/// really the leading run of whatever comes next.
///
/// LOOSE IS THE LAST READING BEFORE GIVING UP. The encoder writes marks that
/// have no reading for their base's form: `iʲ` comes back as `i<y^>` and
/// palatalization is spelled for a consonant. Refusing left the bracket to
/// leak out one raw character at a time, which is strictly worse than reading
/// it as what it plainly says.
fn read_run(
  run: &str,
  form: Form,
  base: Option<&'static Phone>,
  loose: bool,
) -> Option<Vec<&'static Modifier>> {
  if run.is_empty() {
    return None;
  }

  let state = runtime();
  let mut out: Vec<&'static Modifier> = Vec::new();
  let mut at = 0;

  while at < run.len() {
    let (options, length) = state.talk_modifier.match_at(run, at)?;

    let one = match pick_modifier(options, form) {
      Some(one) => Some(one),
      None if loose => options.first().copied(),
      None => None,
    }?;

    if let Some(phone) = base {
      if !modifier_attaches(phone, one) {
        return None;
      }
    }

    out.push(one);
    at += length;
  }

  if out.is_empty() { None } else { Some(out) }
}

/// The body and end of a bracketed modifier run starting at `at`.
///
/// A SCAN, NOT A SEARCH. Searching for the next `>` finds a closing bracket
/// anywhere ahead, so a stray `<` with its match far downstream would swallow
/// every sound between them. Walking forward stops at the first character that
/// cannot be inside a run, so an unclosed `<` fails here and is carried through
/// as an ordinary character.
///
/// Runs do not nest. A `<` inside one is malformed, not a sub-run.
fn run_span(text: &str, at: usize) -> Option<(String, usize)> {
  if text.as_bytes().get(at) != Some(&b'<') {
    return None;
  }

  let mut index = at + 1;

  while index < text.len() {
    match text.as_bytes()[index] {
      b'>' => {
        return if index == at + 1 {
          None
        } else {
          Some((text[at + 1..index].to_string(), index + 1))
        };
      }
      b'<' => return None,
      _ => index += 1,
    }
  }

  None
}

/// How many sounds a binder joins, or `None` when the run is not a binder.
///
/// WHY IT COUNTS BACKWARD. A tie in IPA sits BETWEEN the letters it joins, so
/// `t͡ʃ` says nothing about how far the binding reaches. Naming the count
/// instead makes a doubly-articulated `k͡p` and a three-part cluster equally
/// sayable, and puts the binder after its material.
fn binder_reach(run: &str) -> Option<usize> {
  let (_, rest) = binder_split(run)?;

  let reach = if rest.is_empty() {
    2
  } else {
    rest.parse::<usize>().ok()?
  };

  if reach >= 2 { Some(reach) } else { None }
}

/// A binder run split into the marks before its `B` and the count after it.
///
/// `tx<B>` is a bare affricate and `ts<hB>` an aspirated one, where the `h`
/// belongs to the whole group. The binder is written last because it acts on
/// everything before it.
fn binder_split(run: &str) -> Option<(&str, &str)> {
  let at = run.rfind('B')?;
  let count = &run[at + 1..];

  if !count.chars().all(|one| one.is_ascii_digit()) {
    return None;
  }

  Some((&run[..at], count))
}

/// The marks a binder run carries before its `B`.
fn binder_marks(run: &str) -> &str {
  binder_split(run).map(|(marks, _)| marks).unwrap_or("")
}

/// Several sounds joined into one, as a tie does.
///
/// THE TIE BINDS THE BASES AND THE MARKS HOIST OUT. `t<h>x<B>` is an aspirated
/// affricate and comes back `t͡ʃʰ`, not `tʰ͡ʃ`. A tie says two LETTERS are one
/// segment, so a mark between them breaks the thing it is asserting.
fn bind_sounds(parts: &[Sound]) -> Sound {
  let talk: String = parts.iter().map(|one| one.talk.as_str()).collect();
  let simple: String = parts.iter().map(|one| one.simple.as_str()).collect();

  let bases: Vec<String> = parts
    .iter()
    .map(|one| match one.base {
      Some(base) => base.ipa.clone(),
      None => one.ipa.clone(),
    })
    .collect();

  let marks: String = parts
    .iter()
    .flat_map(|one| one.modifiers.iter().map(|mark| mark.ipa.as_str()))
    .collect();

  let pre: String = parts
    .iter()
    .flat_map(|one| one.pre.iter().map(|mark| mark.ipa.as_str()))
    .collect();

  let ipa = format!("{}{}{}", pre, bases.join("\u{0361}"), marks);
  let count = if parts.len() > 2 {
    parts.len().to_string()
  } else {
    String::new()
  };

  // ONE PAIR OF BRACKETS PER BASE, the same rule the modifiers follow. The
  // binder used to open its own pair, so an aspirated affricate came back
  // `ts<h><B>` with two runs on one sound. It joins the run that is already
  // there instead: `ts<hB>`.
  let bound = if talk.ends_with('>') {
    format!("{}B{count}>", &talk[..talk.len() - 1])
  } else {
    format!("{talk}<B{count}>")
  };

  raw_sound(&SymbolEntry {
    talk: bound,
    ipa,
    simple,
  })
}

/// Split a talk string into sounds. A single starter lookup gives the base (or
/// a symbol); a base then swallows the modifiers that follow it, and the sound
/// is re-emitted in canonical order.
pub fn segment(text: &str) -> Vec<Sound> {
  let state = runtime();
  let mut sounds = Vec::new();
  // Modifiers seen before a base, which modify what FOLLOWS: `h~k` is
  // pre-aspirated, `n~d` prenasalized.
  let mut leading: Vec<&'static Modifier> = Vec::new();

  let mut i = 0;

  while i < text.len() {
    let start_length = state
      .talk_starter
      .match_at(text, i)
      .map(|(_, length)| length)
      .unwrap_or(0);

    // A BRACKETED RUN BEFORE A BASE is the pre-modifiers, `<h>k` for
    // pre-aspirated. Read whole, so the scan below never sees a bracket.
    if start_length == 0 {
      if let Some((body, end)) = run_span(text, i) {
        // A BINDER TIES WHAT CAME BEFORE IT, so it is read here rather than
        // as a modifier on a following base: `tx<B>` is one affricate, and
        // there is no base after the bracket to carry it.
        if let Some(reach) = binder_reach(&body) {
          if sounds.len() >= reach {
            let held: Vec<Sound> = sounds.split_off(sounds.len() - reach);

            sounds.push(bind_sounds(&held));
            i = end;
            continue;
          }
        }

        if let Some((Unit::Phone(next), _)) = state.talk_starter.match_at(text, end)
        {
          let held = read_run(&body, next.form, None, false)
            .or_else(|| read_run(&body, next.form, None, true));

          if let Some(held) = held {
            leading.extend(held);
            i = end;
            continue;
          }
        }
      }
    }

    let Some((&start, length)) = state.talk_starter.match_at(text, i) else {
      // Unknown character: carry it through so nothing is silently
      // dropped.
      let character = text[i..].chars().next().expect("in-bounds character");

      i += character.len_utf8();

      let character = character.to_string();

      sounds.push(raw_sound(&SymbolEntry {
        talk: character.clone(),
        ipa: character.clone(),
        simple: character,
      }));

      continue;
    };

    i += length;

    match start {
      Unit::Phone(phone) => {
        // A BINDER SHARING THE RUN. `ts<hB>` is an aspirated affricate: the
        // marks belong to the base they follow and the `B` ties it to what
        // came before. `bind_sounds` hoists the marks onto the group, so
        // reading them here and binding after gives the same sound.
        if let Some((body, end)) = run_span(text, i) {
          if let Some(reach) = binder_reach(&body) {
            let marks = binder_marks(&body);
            let held = if marks.is_empty() {
              Some(Vec::new())
            } else {
              read_run(marks, phone.form, None, false)
                .or_else(|| read_run(marks, phone.form, None, true))
            };

            if let Some(held) = held {
              sounds.push(make_sound(phone, held, std::mem::take(&mut leading)));

              if sounds.len() >= reach {
                let parts: Vec<Sound> = sounds.split_off(sounds.len() - reach);

                sounds.push(bind_sounds(&parts));
              }

              i = end;
              continue;
            }
          }
        }

        // A BRACKETED RUN AFTER A BASE carries every modifier on it, in one
        // pair. `k<wh>` is labialized and aspirated. The contents are
        // uniquely decodable, so no separator is needed inside.
        if let Some((body, end)) = run_span(text, i) {
          // WHOSE MARK IS IT. A run after a base usually belongs to it, and
          // `attaches` settles the cases where it might not. But the rule
          // alone is too strong: the encoder spells b̥ as `b<v->` and kǃʰ as
          // `k!<h>`, and neither mark attaches to those bases by rule, so
          // refusing outright meant the tokenizer could not read back what
          // its own encoder writes and the brackets spilled out as raw
          // characters.
          //
          // A mark is only handed on when there is somewhere for it to GO:
          // the next sound has to be a phone that accepts the whole run.
          let fits = read_run(&body, phone.form, Some(phone), false);
          let held = fits.clone().or_else(|| {
            read_run(&body, phone.form, None, false)
              .or_else(|| read_run(&body, phone.form, None, true))
          });

          let hand_on = if held.is_some() && fits.is_none() {
            matches!(
              state.talk_starter.match_at(text, end),
              Some((Unit::Phone(next), _))
                if read_run(&body, next.form, Some(next), false).is_some()
            )
          } else {
            false
          };

          if let Some(held) = held {
            if !hand_on {
              sounds.push(make_sound(phone, held, std::mem::take(&mut leading)));
              i = end;
              continue;
            }
          }
        }

        let mods = Vec::new();

        sounds.push(make_sound(phone, mods, std::mem::take(&mut leading)));
      }
      Unit::Symbol(symbol) => sounds.push(raw_sound(symbol)),
      // The starter trie only holds phones and symbols.
      Unit::Modifier(_) => {}
    }
  }

  // A pre-modifier with nothing after it modifies nothing. Carry the
  // spelling through rather than dropping it, so a caller sees the input
  // was incomplete instead of losing it.
  for modifier in leading {
    sounds.push(raw_sound(&SymbolEntry {
      talk: modifier.talk.clone(),
      ipa: modifier.ipa.clone(),
      simple: modifier.simple.clone(),
    }));
  }

  sounds
}
